test/e2e/apimachinery/apply.go

```diff
@@ -0,0 +1,303 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apimachinery
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	"github.com/onsi/ginkgo"
+
+	// ensure libs have a chance to initialize
+	_ "github.com/stretchr/testify/assert"
+)
+
+var _ = SIGDescribe("ServerSideApply", func() {
+	f := framework.NewDefaultFramework("apply")
+
+	var client clientset.Interface
+	var ns string
+
+	ginkgo.BeforeEach(func() {
+		client = f.ClientSet
+		ns = f.Namespace.Name
+	})
+
+	ginkgo.AfterEach(func() {
+	})
+
+	ginkgo.It("PATCH requests with the apply content type will create the object if it does not already exist", func() {
+		{
+			testCases := []struct {
+				resource string
+				name     string
+				body     string
+			}{
+				{
+					resource: "pods",
+					name:     "test-pod",
+					body: `{
+				"apiVersion": "v1",
+				"kind": "Pod",
+				"metadata": {
+					"name": "test-pod"
+				},
+				"spec": {
+					"containers": [{
+						"name":  "test-container",
+						"image": "test-image"
+					}]
+				}
+			}`,
+				}, {
+					resource: "services",
+					name:     "test-svc",
+					body: `{
+				"apiVersion": "v1",
+				"kind": "Service",
+				"metadata": {
+					"name": "test-svc"
+				},
+				"spec": {
+					"ports": [{
+						"port": 8080,
+						"protocol": "UDP"
+					}]
+				}
+			}`,
+				},
+			}
+
+			for _, tc := range testCases {
+				_, err := client.CoreV1().RESTClient().Patch(types.ApplyPatchType).
+					Namespace(ns).
+					Resource(tc.resource).
+					Name(tc.name).
+					Param("fieldManager", "apply_test").
+					Body([]byte(tc.body)).
+					Do(context.TODO()).
+					Get()
+				if err != nil {
+					framework.Failf("Failed to create object using Apply patch: %v", err)
+				}
+
+				_, err = client.CoreV1().RESTClient().Get().Namespace(ns).Resource(tc.resource).Name(tc.name).Do(context.TODO()).Get()
+				if err != nil {
+					framework.Failf("Failed to retrieve object: %v", err)
+				}
+
+				// Test that we can re apply with a different field manager and don't get conflicts
+				_, err = client.CoreV1().RESTClient().Patch(types.ApplyPatchType).
```

Comment:
should this be inspecting the result for some expected state? do we expect two field managers that co-own these fields, or that apply_test_2 owns no fields (since it wasn't changing any state)?

Reply:
cc @apelisse, I think the two managers will co own the fields?

From a conformance perspective, managed fields are opaque and shouldn't be something users need to directly deal with. Wouldn't it make more sense to try to perform operations after a specific state is reached instead of directly inspecting the managed fields? (Eg: What happens when one of the managers applies something afterwards?) I added another test for the case when two managers own a field. When one manager drops ownership of that field, the field should not be deleted.

Reply:
> cc @apelisse, I think the two managers will co own the fields?

Yes.

> From a conformance perspective, managed fields are opaque

They are part of the API at this point, they can't be treated as opaque.

Reply:
ack, added a check for the managedFields. It compares the stringified managed fields with a const string. Not sure if that's the ideal way to do it?